[4-6단계 - 방탈출 사용자 예약] 부기(이창의) 미션 제출합니다.

changuii opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->

안녕하세요 로키! 피드백 항상 감사합니다! :)

## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [ ] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
메서드의 길이가 많이 길어졌고, indent도 지키지 못한 부분들이 있다.

## 웹 개발 경험 체크
🌱 레벨2 시작 이전에, 웹 애플리케이션 개발 경험이 있나요?
- [ ] 아니오. 전혀 경험이 없습니다.
- [ ] 예. 예제 코드를 따라하거나 클론 코딩을 해봤습니다. (ex. 강의를 보고 따라함)
- [ ] 예. 웹 애플리케이션을 만들었지만 로컬에서만 실행해봤습니다.
- [x] 예. 웹 애플리케이션을 만들고 배포해봤습니다. (ex. 클라우드 혹은 리모트 서버에 배포)

## 베이스 코드 선택 체크
- [ ] 이전 미션의 내 코드에서 시작 
- [x] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->

현재 테스트를 열심히 작성하긴 했지만 스스로 완벽하게 모든 부분을 찾아낼 수 있는 테스트 코드는 아니라고 생각합니다. 이에 대해서 피드백 부탁드립니다.
- 테스트가 너무 복잡해지는데 어떻게 해결하면 좋을지 키워드 부탁드립니다!!

#### 개인적인 궁금증
1. 특정 필터링 조건이 존재하는 테스트 코드를 작성하는 것이 어려운데, 팁이나 학습 키워드가 있는지 궁금합니다.!
2. 레벨 2로 오면서 많은 고민을 하기 보다는 스프링이라는 도구를 사용하는 방법에 대해서 학습하고 있다고 느껴지는데 어떤 부분을 집중적으로 고민해보아야 하고, 학습해야하는지 잘 모르겠습니다. 로키는 이 시절에 어떤 부분을 주로 학습하셨는지 궁금합니다. 이렇게 도구를 학습한다고 생각하고 외우고 익히는데 집중하는게 올바른 방향인지 로키의 피드백이 궁금합니다.
3. 개발을 하면서 의문점이 많이 생겼습니다. 이를 테면, 객체는 참조를 통해 다른 객체들과 관계를 맺지만, 데이터베이스는 외래키 제약조건을 통해 관계를 맺고 있었습니다. 때문에, 레벨 1때 배웠던 객체의 관점에서 생각해보았을 때 null이 존재하는 것이 어색하고 위험하다고 느껴지지만, 데이터베이스에 생성하기 전에는 id가 없는게 당연하기 때문에, 어쩔 수 없는 부분들이 생겼고 계속 이렇게 타협하다보니 점점 방어적인 코드를 작성하는게 어려워졌습니다.  그래서 생각해본 해결책은 테이블을 나타내는 객체를 service-repository간에 추가로 만들고, service layer에서 이 테이블을 나타내는 객체를 조합해 도메인 객체를 생성하여 사용하는 ? 구조가 떠오르긴 했습니다. 하지만 그만큼 많은 코드와 복잡성이 따라온다고 느껴졌습니다. 여기서 스스로 확신이 생기지 않아 구글링과 ai를 활용해서 찾다보니 DDD에 도달했는데 지금 이 설계를 학습하는게 올바른 방향성인지 궁금합니다!!
  - 계속 리팩터링을 하면서 문제가 된다고 느낀 가장 큰 부분은 데이터베이스 관점에서는 모든 테이블을 볼 수 있고(캡슐화라는 개념이 없음) 이를 연관된 repository에서 접근하다보니, 하나의 도메인에 수정사항이 생기면 많은 repository에서 변경이 전파되는 문제가 존재했고 무언가 repository에서 접근할 수 있는 데이터의 영역을 제한(객체지향의 캡슐화처럼 접근할 수 있는 테이블의 제한)해야할 필요성을 느꼈지만 어떻게 풀어가야할지 잘 모르겠어서,, 이에 대한 키워드도 부탁드립니다!


### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

changuii committed:
fix: findAllByThemeIdAndDate() query에 parameter를 넣지 않던 문제 해결

changuii committed:
remove: printStackTrace() 제거

changuii committed:
test: ReservationTime 테스트 추가

changuii committed:
feat: 4단계 로그인 페이지 활성화

changuii committed:
test: 유저 페이지 테스트 추가

changuii committed:
feat: 회원가입 페이지 추가

changuii committed:
test: 회원가입 페이지 테스트

changuii committed:
feat: member 테이블 추가

changuii committed:
feat: 멤버 생성 기능 추가

changuii committed:
test: 멤버 생성 기능에 대한 테스트 추가

changuii committed:
chore: jwt dependency 추가

changuii committed:
feat: 로그인시 토큰 발급 기능 추가

changuii committed:
test: 로그인 기능 테스트 추가

changuii committed:
test: member 테스트 추가

changuii committed:
feat: 토큰 기반 유저 정보 조회 기능 추가

changuii committed:
test: /login/check 기능 테스트

changuii committed:
feat: ArgumentResolver 추가

changuii committed:
test: AuthService 테스트 변경

changuii committed:
feat: 토큰 기반으로, 예약을 생성한다.

changuii committed:
test: 토큰 기반으로, 예약을 생성 테스트 수정 및 추가

changuii committed:
remove: printStackTrace() 제거

changuii committed:
feat: member 전체 조회 기능 추가

changuii committed:
test: member 전체 조회 기능 테스트 추가

changuii committed:
chore: 초기 테이터 삽입

changuii committed:
feat: 관리자 예약 생성 기능 추가

changuii committed:
test: 관리자 예약 생성 기능 테스트 추가

changuii committed:
feat: Member 테이블에 role 컬럼 추가

changuii committed:
test: Member 테이블에 role 컬럼 추가로 변경된 테스트 변경

changuii committed:
feat: /admin/** path에는 오직 admin 로그인 사용자만 접근할 수 있도록 구현

changuii committed:
test: admin 접근 구역에 ADMIN 계정 쿠키 추가

changuii committed:
feat: 멤버 id, 테마 id, 날짜 범위를 기준으로 에약을 조회하는 기능 추가

changuii committed:
test: 멤버 id, 테마 id, 날짜 범위를 기준으로 에약을 조회하는 기능 테스트 추가

Rok93 reviewed (changes requested):
부기 4~6단계도 빠르게 구현해주셨네요. 😃
다만 현재 깨지는 테스트가 꽤 있는데, 한번 확인해서 통과하도록 수정해서 다시 리뷰요청 부탁드릴게요. 🙏

changuii committed:
remove: 불필요한 scheme.sql 제거

changuii committed:
test: member 테이블의 튜플을 제거하기 전에 외래키 제약조건으로 연관된 reservation 테이블의 튜플을 먼저 제거

- 실행 순서에 따라 외래키 제약조건 발생 가능성 제거

Rok93 commented on src/main/java/roomescape/auth/AuthController.java:
```diff
@@ -0,0 +1,49 @@
+package roomescape.auth;
```
질문 남겨주신 부분에 대해서는 여기서 하나씩 제 생각을 남겨볼게요. 

1. 특정 필터링 조건이 존재하는 테스트 코드를 작성하는 것이 어려운데, 팁이나 학습 키워드가 있는지 궁금합니다.!

> 이 부분은 조금만 더 추가로 설명해주실 수 있을까요? 필터링 조건이 존재하는 테스트 코드라는게 어떤 부분에서 어려움을 느끼셨는지를 부연설명해주시면 조금 더 유의미한 답변을 드릴 수 있지 않을까 싶어요. 🙏

Rok93 replied:
2. 레벨 2로 오면서 많은 고민을 하기 보다는 스프링이라는 도구를 사용하는 방법에 대해서 학습하고 있다고 느껴지는데 어떤 부분을 집중적으로 고민해보아야 하고, 학습해야하는지 잘 모르겠습니다. 로키는 이 시절에 어떤 부분을 주로 학습하셨는지 궁금합니다. 이렇게 도구를 학습한다고 생각하고 외우고 익히는데 집중하는게 올바른 방향인지 로키의 피드백이 궁금합니다.

> 레벨2는 스프링에 대한 학습을 하는 단계라고 생각해서 스프링이라는 도구를 학습하고 사용 방법에 대해서 학습한다고 느껴지는게 당연하다고 생각해요. 다만 도구를 학습하기위해서 단순히 `외우고 익히는데 집중`하는 것은 그렇게 좋은 방법은 아니라고 생각해요. 😃
이론적인 학습과 더불어서 가능하다면 학습 테스트를 작성해보고 더 나아가서 디버깅 모드로 실제로 이론적으로 학습한 것처럼 flow가 흘러가는지를 눈으로 직접 보는 과정들이있으면 레벨2를 정말 슬기롭게해나가고있다고 생각해요. 추가적으로 화룡 정점으로 TIL, 블로그 등을 통해서 Deep Dive하게 학습한 내용들을 기록해둔다면 더더욱 💯점 만점일 것 같네요. 🤗 
(이렇게 학습한 내용들은 나중에 면접을 보실 때도 도움이 많이 되실거라 생각해요. 😃)

Rok93 replied:
3. 개발을 하면서 의문점이 많이 생겼습니다. 이를 테면, 객체는 참조를 통해 다른 객체들과 관계를 맺지만, 데이터베이스는 외래키 제약조건을 통해 관계를 맺고 있었습니다. 때문에, 레벨 1때 배웠던 객체의 관점에서 생각해보았을 때 null이 존재하는 것이 어색하고 위험하다고 느껴지지만, 데이터베이스에 생성하기 전에는 id가 없는게 당연하기 때문에, 어쩔 수 없는 부분들이 생겼고 계속 이렇게 타협하다보니 점점 방어적인 코드를 작성하는게 어려워졌습니다. 그래서 생각해본 해결책은 `테이블을 나타내는 객체를 service-repository간에 추가로 만들고, service layer에서 이 테이블을 나타내는 객체를 조합해 도메인 객체를 생성하여 사용하는 ? 구조`가 떠오르긴 했습니다. 하지만 그만큼 많은 코드와 복잡성이 따라온다고 느껴졌습니다. 여기서 스스로 확신이 생기지 않아 구글링과 ai를 활용해서 찾다보니 DDD에 도달했는데 지금 이 설계를 학습하는게 올바른 방향성인지 궁금합니다!!

말씀하신 관점을 찾다보면 DDD가 나올 수도 있을 것 같다...? 라는 생각도 들긴하지만, 질문 주신 내용을 다시 되짚어보자면
결과적으로 객체지향 vs RDB 간의 패러다임 불일치에의해 발생하는 현상이라고 생각해요. DDD가 이런 문제를 해결해줄거라고 생각되지는 않는 것 같아요. 👀 개인적인 생각에서는 결국 도메인 객체가 곧 테이블과 매칭이되는 구조에서 발생하는 문제로 보여요. 결과적으로 이런 두 용도의 객체를 분리하여 순수 도메인 객체 vs (테이블과 매칭되는) Entity 객체 로 나눈다면 이런 문제점을 어느정도 해소할 수 있지 않을까 싶어요. 🤔 (아마 이 방식이 추가로  만든다는 테이블을 나타내는 객체가 아닐까 싶기도하네요. 👀)

`그만큼 많은 코드와 복잡성이 따라온다고 느껴졌습니다` << 이 부분은 해보지도 않으셨는데, 감이 잡히시나보군요...? 🤩 
말씀하신 것처럼 말씀하신 문제점들이 싫고 이를 지키기위해 추가로 도입하는 것들로인해서 시스템의 복잡도가 더 커지는 경우가있고, 결국 트레이드 오프를 따져서 더 낫다고 생각하는 방향으로 의사결정을하게되는데요. 
부기도 이런 트레이드 오프의 순간에 많이 놓이게될텐데요. 이번 미션에서도 이 트레이드 오프를 잘 따져보고 선택을 해보면 좋을 것 같아요. 😃

질문에서 `테이블을 나타내는 객체를 service-repository간에 추가로 만들고, service layer에서 이 테이블을 나타내는 객체를 조합해 도메인 객체를 생성하여 사용하는 ? 구조`의 내용을 도전해보시는건 어떨까요? 😃  사실 지금은 텍스트로만 부기의 의도를 듣다보니 조금 제가 생각하는 것과 동일한 생각을하신건지 잘 모르겠어서요. 

사실 저도 이 방식은 말로만 들어봤지 직접 시도해본 적은 없어서, 부기가 시도하신다면 어떻게 진행하실지가 조금 궁금하긴하네요. 😁
만약 도전(?)하신다면 저도 추가로 더 알아보도록할게요. 😃 (아..! 물론 트레이드 오프 먼저 따져보시면 좋을 것 같습니다)
다만 다음 단계에서.... 혹시나 부기의 코드를 베이스로 진행하게된다면 같이 진행하시는 페어분은 많이 당황하실 수도 있겠네요. 😅 

![image](https://github.com/user-attachments/assets/74680afe-102d-490d-a630-04e366a131c4)

Rok93 replied:
4. 계속 리팩터링을 하면서 문제가 된다고 느낀 가장 큰 부분은 데이터베이스 관점에서는 모든 테이블을 볼 수 있고(캡슐화라는 개념이 없음) 이를 연관된 repository에서 접근하다보니, 하나의 도메인에 수정사항이 생기면 많은 repository에서 변경이 전파되는 문제가 존재했고 무언가 repository에서 접근할 수 있는 데이터의 영역을 제한(객체지향의 캡슐화처럼 접근할 수 있는 테이블의 제한)해야할 필요성을 느꼈지만 어떻게 풀어가야할지 잘 모르겠어서,, 이에 대한 키워드도 부탁드립니다!

Repository가 매칭되는 Entity만 수정하는 것이 맞다고 생각해요. 
한번에 여러 Entity에 수정이 발생한다면, 이건 테이블 구조를 잘못 분리한게 아닐까...? 라고 생각되기도하는데요. 🤔
테이블 구조를 잘 정하는 것이 중요하다고 생각하는데... 문제되는 케이스가 어떤건지가 궁금하네요. 

그런데 `하나의 도메인에 수정사항이 생기면 많은 repository에서 변경이 전파되는 문제가 존재했고`라고 하셨는데, 방탈출 사용자 예약 미션에서 이런 케이스가 존재했나요? 🤔 기능중에 뭔가 생각나는 케이스가 없는데, 어떤 케이스가 있었는지 공유해주실 수 있을까요? 

말씀하시는 내용들과 딱 맞는 키워드는 사실 생각나는게 없는데요. 👀
DDD 학습을 이미 하신듯하니 가볍게 DDD 관련 키워드를 드려보자면 **애그리거트**, **애그리거트 루트** 같은 키워드가 말씀하신 내용과 조금은 연관이될 수도 있을 것 같아요.  개인적인 생각으로는 DDD 개념은 아직은 미지의 영역일거라 생각해서 가볍게 키워드 정도만 찾아보시되 잘 이해가 안되신다면 너무 시간 많이 안쏟는걸 추천드려요. 😃

changuii replied:
1. 이 부분은 조금만 더 추가로 설명해주실 수 있을까요? 필터링 조건이 존재하는 테스트 코드라는게 어떤 부분에서 어려움을 느끼셨는지를 부연설명해주시면 조금 더 유의미한 답변을 드릴 수 있지 않을까 싶어요. 🙏

---

ReservationTimeController.findAllAvailableTimes() 메서드와 같은 부분은 여러가지 필터링 조건이 존재하는데 이런 경우에 어떤 경우를 테스트해야할지 감이 잡히지 않아서,, 테스트케이스를 작성하기 어려웠습니다.
- themeId, date를 기준으로 테스트를 작성해야하는데 
	- 테마를 여러개 넣어두고 테마가 특정 테마 id를 가진 값이 나오는지 검증
	- date를 여러개 넣어두고 특정 date를 가진 값이 나오는지 검증
	- 테마와 date를 여러개 넣어두고 테마 id를 가진 값과 date를 가진 값이 나오는지 검증
위의 테스트 케이스를 만드는 것처럼 여러가지 조합을 테스트하는게 올바른 방향인지 잘 모르겠습니다.

추가로, ThemeService의 findTopRankThemes() 메서드에서  ThemeRepository.findAllOrderByRank() 메서드를 사용하고 있는데, Fake에서 이를 구현하려고하니 Theme의 값이 아니라 Reservation 테이블의 Date를 기준으로 정렬하고 있어 이를 실제 ArrayList를 사용하는 Fake에서 구현하기 어려웠습니다.
제가 메서드의 위치를 잘못 고려한 탓일까요? 아니면 ArrayList로 구현한 Fake의 한계인지 잘 모르겠습니다.

changuii replied:
#### 2. 레벨2는 스프링에 대한 학습을 하는 단계라고 생각해서 스프링이라는 도구를 학습하고 사용 방법에 대해서 학습한다고 느껴지는게 당연하다고 생각해요. 다만 도구를 학습하기위해서 단순히 외우고 익히는데 집중하는 것은 그렇게 좋은 방법은 아니라고 생각해요. 😃
이론적인 학습과 더불어서 가능하다면 학습 테스트를 작성해보고 더 나아가서 디버깅 모드로 실제로 이론적으로 학습한 것처럼 flow가 흘러가는지를 눈으로 직접 보는 과정들이있으면 레벨2를 정말 슬기롭게해나가고있다고 생각해요. 추가적으로 화룡 정점으로 TIL, 블로그 등을 통해서 Deep Dive하게 학습한 내용들을 기록해둔다면 더더욱 💯점 만점일 것 같네요. 🤗
(이렇게 학습한 내용들은 나중에 면접을 보실 때도 도움이 많이 되실거라 생각해요. 😃)

---

1. 감사합니다 로키! 혹시 괜찮으시면 로키만의 학습 방법이 궁금합니다! 🤔
	- 말씀하신 것처럼, 학습한 내용을 학습 테스트를 통해 눈으로 확인해보고 이 내용들을 블로그나, 어딘가에 기록하는 방식으로 학습하셨을까요?
2. 레벨1을 보내면서 스스로 학습 방법을 최대한 깊은 단계에서 직접 눈으로 확인해보고 이해하려고 마음을 먹었으나,, 레벨2에서 스프링으로 넘어오면서 이런 학습 방법이 많은 시간이 소요되고 코드 레벨에서 이해하기 힘든 부분들도 많이 있었습니다.
	- 로키는 깊이있는 학습을 하셨는지 아니면 도구의 동작 방식을 이해하기 위해 깊이있는 학습을 하셨는지 궁금합니다.
	- 저는 호기심에 여러 부분들을 눈으로 직접 확인(어떻게 코드상에서 톰캣과 스프링이 연결될까?와 같은 깊이)해보고 싶은데 스프링이라는 도구를 사용하기 위해 학습한다면, 사실 동작 방식까지 이해할 필요는 없다고 최근에 점점 느껴지더라구요! 로키는 이런 고민 하셨는지 궁금합니다.!!!

특히, 레벨1에서 코드에 의도를 담기 위해 노력한 부분들이 많았는데 레벨2에서는 어디서 의도를 나타내야할지 잘 모르겠더라구요! 이전에  제가 2번째 고민에서 말씀드린 것과 동일하게 도구를 학습하는 과정이라 그런거겠죠? :)

감사합니다 로키!

changuii replied:
##### 3. 테이블을 나타내는 객체를 service-repository간에 추가로 만들고, service layer에서 이 테이블을 나타내는 객체를 조합해 도메인 객체를 생성하여 사용하는 ? 구조의 내용을 도전해보시는건 어떨까요? 😃 

감사합니다. 로키!
시간이 된다면 시도해보겠습니다.

changuii replied:
##### 4. Repository가 매칭되는 Entity만 수정하는 것이 맞다고 생각해요.
한번에 여러 Entity에 수정이 발생한다면, 이건 테이블 구조를 잘못 분리한게 아닐까...? 라고 생각되기도하는데요. 🤔
테이블 구조를 잘 정하는 것이 중요하다고 생각하는데... 문제되는 케이스가 어떤건지가 궁금하네요.

> 그런데 하나의 도메인에 수정사항이 생기면 많은 repository에서 변경이 전파되는 문제가 존재했고라고 하셨는데, 방탈출 사용자 예약 미션에서 이런 케이스가 존재했나요? 🤔 기능중에 뭔가 생각나는 케이스가 없는데, 어떤 케이스가 있었는지 공유해주실 수 있을까요?

---


Reservation의 경우, ReservationTime, Theme, Member와의 관계를 가지고 있었고 현재 제가 작성한 코드에서 만약 이런 연관된 테이블의 변경이 발생할 경우, ReservationRepository의 조회 쿼리가 모두 수정되는 문제가 존재했습니다! 
- ReservationRepository에서 완성된 Reservation을 반환하기 때문에 쿼리에서 다른 테이블들의 컬럼들에 접근하고 있습니다.!

Rok93 replied:
> ReservationTimeController.findAllAvailableTimes() 메서드와 같은 부분은 여러가지 필터링 조건이 존재하는데 이런 경우에 어떤 경우를 테스트해야할지 감이 잡히지 않아서,, 테스트케이스를 작성하기 어려웠습니다.
> themeId, date를 기준으로 테스트를 작성해야하는데
테마를 여러개 넣어두고 테마가 특정 테마 id를 가진 값이 나오는지 검증
date를 여러개 넣어두고 특정 date를 가진 값이 나오는지 검증
테마와 date를 여러개 넣어두고 테마 id를 가진 값과 date를 가진 값이 나오는지 검증
위의 테스트 케이스를 만드는 것처럼 여러가지 조합을 테스트하는게 올바른 방향인지 잘 모르겠습니다.

얼마나 꼼꼼하게 테스트하냐의 문제라고 생각해요. 
아무래도 필터 로직의 경우 5개의 필터 요소가 있다면 5가지 필터를 ON or OFF 하는 조합의 수만큼의 테스트 케이스가 발생할 수 있는데요. 😅  이건 너무 많다는 생각이드는 것 같아요. 개인적으로 저는 각 필터 요소를 하나씩만 테스트해서 정상 동작하는지 확인해도 충분하다고 생각해요. (날짜 범위 같은 경우에는 from, to 조건을 같이해서 범위 조건으로 테스트해봐도 좋을 것 같구요)



> 추가로, ThemeService의 findTopRankThemes() 메서드에서 ThemeRepository.findAllOrderByRank() 메서드를 사용하고 있는데, Fake에서 이를 구현하려고하니 Theme의 값이 아니라 Reservation 테이블의 Date를 기준으로 정렬하고 있어 이를 실제 ArrayList를 사용하는 Fake에서 구현하기 어려웠습니다.
제가 메서드의 위치를 잘못 고려한 탓일까요? 아니면 ArrayList로 구현한 Fake의 한계인지 잘 모르겠습니다.


`findAllOrderByRank`를 ThemeRepository 에서 수행하고있지만 실제 동작하는 쿼리를 생각해봤을 때는 실제 반환되는 결과는 List<Theme>지만 실제로는 reservation의 집계가 더 중요한 쿼리라고 생각해요. 
ThemeRepository 자체에서는 reservation에 대한 

만약 이 findAllOrderByRank 쿼리를 reservationRepository에 위치시킨다면 아래와 같이 Fake 객체를 구현할 수 있을 것 같아요. 👀

```java
    @Override
    public List<Theme> findAllOrderByRank(final LocalDate from, final LocalDate to, final int size) {
        return new ArrayList<>(reservations).stream()
            .filter(it -> it.getDate().isAfter(from) && it.getDate().isBefore(to))
            .collect(Collectors.groupingBy(Function.identity(), Collectors.counting()))
            .entrySet().stream()
            .sorted((o1, o2) -> o2.getValue().compareTo(o1.getValue()))
            .limit(size)
            .map(Map.Entry::getKey)
            .toList();
    }
```


###

Rok93 replied:
> 감사합니다 로키! 혹시 괜찮으시면 로키만의 학습 방법이 궁금합니다! 🤔
말씀하신 것처럼, 학습한 내용을 학습 테스트를 통해 눈으로 확인해보고 이 내용들을 블로그나, 어딘가에 기록하는 방식으로 학습하셨을까요?

TIL까지는 아니더라도 WIL 등을 꾸준히 쓰려고했고 그 외에도 블로그를 꾸준하게 쓰려고 노력했었어요. 
하지만 지금 생각해보면 아쉬움이 많이 남았던 것 같아요. 
단순히 학습한 내용에대한 요약 정도의 글밖에 쓰지 않았었거든요. 🥲 
하지만 뭐가됐건간에 결국 공부한 내용은 언젠가 잊어먹기 마련이에요. 
어딘가에 기록해두지 않는다면 기껏 고생해서 학습한 내용이 금방 휘발될 수 있기 때문에 꼭 기록해둔 뒤, 나중에 까먹었을 때 나만의 방식으로 메모해둔 TIL, 블로그 등을 보면서 쉽게 복습하는 것이 도움이 많이 됐던 것 같아요. 😃 (분명 공부했던건데 기억이 안나서 검색해서 찾았는데 다른 블로그 글은 잘 이해가 안됐지만 우연히 이전에 작성했던 제 블로그 글을 쭉 읽은게 제일 이해가 잘됐었던...  같은 사람이다보니? 같은 공부를하면서 겪는 어려운 포인트가 비슷한 경우가 많았던 것 같아요. 😁)

> 레벨1을 보내면서 스스로 학습 방법을 최대한 깊은 단계에서 직접 눈으로 확인해보고 이해하려고 마음을 먹었으나,, 레벨2에서 스프링으로 넘어오면서 이런 학습 방법이 많은 시간이 소요되고 코드 레벨에서 이해하기 힘든 부분들도 많이 있었습니다.
로키는 깊이있는 학습을 하셨는지 아니면 도구의 동작 방식을 이해하기 위해 깊이있는 학습을 하셨는지 궁금합니다.
저는 호기심에 여러 부분들을 눈으로 직접 확인(어떻게 코드상에서 톰캣과 스프링이 연결될까?와 같은 깊이)해보고 싶은데 스프링이라는 도구를 사용하기 위해 학습한다면, 사실 동작 방식까지 이해할 필요는 없다고 최근에 점점 느껴지더라구요! 로키는 이런 고민 하셨는지 궁금합니다.!!!
특히, 레벨1에서 코드에 의도를 담기 위해 노력한 부분들이 많았는데 레벨2에서는 어디서 의도를 나타내야할지 잘 모르겠더라구요! 이전에 제가 2번째 고민에서 말씀드린 것과 동일하게 도구를 학습하는 과정이라 그런거겠죠? :)

저도 레벨 2를 학습하던 시절에는 단순히 주어진 미션을 잘 구현하는데에 집중했던 것 같아요. 
생각보다 저의 학습 수준에서는 버거운 내용들이 많았었거든요. 😅
질문주신 방향(지금보니 두 선택지 다 비슷해보이는데....? 아마 문맥상 다른 하나는 이 의미가 아닐까 싶어 알잘딱깔센하게 적어볼게요)  중에서 콕 짚어서 말씀드리자면 `스프링이라는 도구를 사용하기 위해 학습`했던 것 같아요. 😃 


> 저는 호기심에 여러 부분들을 눈으로 직접 확인(어떻게 코드상에서 톰캣과 스프링이 연결될까?와 같은 깊이)해보고 싶은데 스프링이라는 도구를 사용하기 위해 학습한다면, 사실 동작 방식까지 이해할 필요는 없다고 최근에 점점 느껴지더라구요! 로키는 이런 고민 하셨는지 궁금합니다.!!!

초기에는 이렇게 생각하실 수도 있다고 생각해요. 
하지만... 이런저런 요구사항을 구현하면서 혹은 에러 상황을 디버깅하다보면 이런 동작원리들을 아냐는 꽤 많은 차이를 불러오게됩니다. 
예를들자면 어떤 장애가 발생해서 디버깅을 할 때, 이런 동작원리를 잘 알게되면 문제 범위를 좁혀서 빠르게 접근해나갈 수 있게되니 더 빠르게 문제해결을 할 수 있게되는 것이죠. 

지금까지 이렇게 말하는 것만 봤을 때는 마치 도구의 동작 방식을 이해하기 위해 깊이있는 학습을 권장하는 것 처럼 들리겠지만, 그건 아닙니다. 순위를 정하고 1순위 목표를 달성한 뒤 2순위 목표를 도전하는 식으로 접근을 하는 것이 좋다고 생각해요. 

코드 품질을 너무 신경쓰다가 마감 기한을 지키지 못하는 개발자처럼 어찌됐건간에 기한내에 1순위 목표를 지키는 것이 최우선이고 1순위 목표를 달성한다면 2순위로 넘어가는 것이 맞다고 생각해요. 

지극히 주관적인 생각이지만 ...
레벨2의 1순위 목표는 `스프링이라는 도구를 사용하기 위한 학습`이라고 생각해요. 
1순위를 빠르게 달성한 후, 2순위의 디테일함을 채워나가는 것이 제가 생각했을 때는 올바른 전략(?)이 아닐까 생각해요.  😃

Rok93 replied:
> Reservation의 경우, ReservationTime, Theme, Member와의 관계를 가지고 있었고 현재 제가 작성한 코드에서 만약 이런 연관된 테이블의 변경이 발생할 경우, ReservationRepository의 조회 쿼리가 모두 수정되는 문제가 존재했습니다!
ReservationRepository에서 완성된 Reservation을 반환하기 때문에 쿼리에서 다른 테이블들의 컬럼들에 접근하고 있습니다.!

아쉽게도 연관관계를 가진 Entity는 이런 상황이 어쩔 수 없지 않나...? 라고 생각해요. 🤔
이런 관점이 아니라면 Repository가 다른 Repository에 의존하는 방식도 고민해볼 수 있지만
개인적으로 같은 계층끼리는 의존하지 않도록 계층을 유지하는 것을 지향하기 때문에 지금의 방식 외에는 딱히 생각나는 방식이 없는 것 같아요. 👀

Rok93 commented on src/main/java/roomescape/auth/dto/LoginMember.java:
```diff
@@ -0,0 +1,4 @@
+package roomescape.auth.dto;
+
+public record LoginMember(Long id, String name, String email, String role) {
```
role을 문자열이아닌 enum으로 관리해보면 어떨까요? 😃

changuii replied:
피드백 감사합니다. :) enum으로 관리하는게 더 좋은 것 같습니다.! 
반영하였습니다.

Rok93 commented on src/main/java/roomescape/auth/AuthorizationAdminInterceptor.java:
```diff
@@ -0,0 +1,49 @@
+package roomescape.auth;
```
auth 패키지 하위의 요소들 중에서는 설정과 관련된 요소도 많이있는 것으로 보이는데요. 👀
이런 케이스는 config 패키지 하위에 들어가는게 맞지 않을까 생각하는데 부기는 어떻게 생각하시나요? 😃

changuii replied:
로키가 config 패키지 하위에 들어가는게 맞다고 판단하신 이유가 궁금합니다,.!

### 나의 생각 및 판단 
인증 및 인가 처리를 위해 interceptor와 Resolver를 구현하였고 이 클래스들은 인증 및 인가와 관련있기 때문에 auth 패키지에 있는 것이 자연스럽다고 판단하여 이렇게 구현하였습니다.
실제로 config의 WebMvcConfiguration은 실제로 이런 클래스들을 활용해 스프링에게 설정(제공)해주고 있기 때문에 해당 패키지에 두었습니다.

Rok93 replied:
먼저 조금 생략한 내용이있어 추가로 말씀을드리자면 
해당 피드백은 auth 패키지 하위의 모든 클래스를두고 말씀드린 부분은 아닙니다. 😅
AuthController, AuthService 같은 경우는 제외하고 말씀드리는거에요! 🙏

그리고 auth 패키지 하위에 이런 구현체들이 있는 것을두고 말씀드린게아니라, auth 패키지가 config 패키지와 동등한 레벨로 위치한 것에 대한 코멘트였어요! 

뭔가 엄청난 논리가 있어서 말씀드렸던 것은 아니고, 저는 결국 interceptory와 Resolver 모두 `WebMvcConfiguration`에 등록해주는 설정(=configuration) 요소들이라고 생각하기 떄문에 config 하위에 위치하는 것이 맞다고생각하고, config 하위에 별도로 auth와 같은 패키지를 두고 해당 요소들을 위치시키는 것이 (config에 등록하는 요소들을 config 하위에 위치시킨다는 관점에서) 조금 더 패키지 구조를 직관적으로 파악할 수 있지 않을까...? 라고 생각해요.

changuii replied:
interceptor와 resolver가 auth 패키지 내의 다른 클래스들과 어떤 것들이 다를까 고민해보았습니다.
interceptor와 resolver는 계층형 아키텍처인 controller - service - repository 구조에 포함되지 않고 스프링이 제공하는 DispatcherServlet에게 제공하는 기능이었습니다.
그렇다면 각각의 패키지들이 이런 계층형 아키텍처에서의 도메인을 기준으로 나누었다면 resolver와 interceptor는 api의 주요 기능이라기 보다는 공통으로 제공할 수 있는 부가적인 기능이라는 생각이 들었습니다.
auth라는 도메인만을 위한 기능이라기 보다는 예를 들어 현재 만들어진 interceptor는 특정 api들에게 ADMIN만 접근할 수 있도록 제어하는 역할을 하고, resolver는 특정 api들에게 토큰으로 유저 정보를 dto로 매핑해주는 역할을 하고 있습니다.
이렇게 작성하고 보니 auth라는 도메인에 대한 역할 보다는, api들이 공통으로 사용하는 부가기능에 가깝지 않을까? 라는 생각이 들었습니다. 
반영하겠습니다!

Rok93 commented on src/main/java/roomescape/auth/AuthorizationAdminInterceptor.java:
```diff
@@ -0,0 +1,49 @@
+        return true;
+    }
+
+    private String extractToken(final HttpServletRequest request) {
+        return Arrays.stream(request.getCookies())
+                .filter(cookie -> Objects.equals(cookie.getName(), TOKEN_NAME))
+                .map(Cookie::getValue)
+                .findAny()
+                .orElseThrow(() -> new AuthNotExistsCookieException());
+    }
```
토끼가 우려하고있는 것처럼 쿠키가 없는 경우 인터셉터가 동작하게되면 NPE가 발생하고있네요. 👀 
null에대한 처리해주시면 좋을 것 같아요. 😃

changuii replied:
토큰이라는 키를 가진 쿠키가 존재하지 않는 경우는 생각했으나,, 쿠키가 존재하지 않을 때 null 값이 나오는지 몰랐습니다. 
토큰을 추출하기 전에 검증 로직을 추가하여 해결했습니다.
피드백 감사합니다. :) 😃

Rok93 commented on src/main/java/roomescape/reservation/AdminReservationController.java:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    @GetMapping
+    public ResponseEntity<List<ReservationResponse>> readAllByMemberAndThemeAndDateRange(
```
현재 필터 조건에 하나라도 값이 셋팅되지 않는다면 예약 검색이 되지 않는 것 같은데요. 🤔
실제 웹 사이트에서 필터링 검색들을 떠올려봤을 때, 꼭 모든 필터 값을 지정하지 않아도 검색할 수 있지 않나요? 😃 

<img width="1774" alt="image" src="https://github.com/user-attachments/assets/7b31ec84-ae32-48c2-88ca-1af930751e0d" />

changuii replied:
### 나의 판단 
모든 필터를 선택해야 검색이 가능하도록 해야겠다 라고 판단하여 이렇게 구현하였습니다.
- 특히, 한번 설정한 값을 되돌릴 수 없었기 때문에 모두 존재해야 검색이 가능하겠구나 라고 판단하였습니다.
- 또한, 필터를 각각의 분기 처리를 통해 동적인 쿼리를 만들어야 한다고 생각했고, 매우 많은 분기문이 생길 것으로 예상하여 이런 의도가 아닐 것이라 판단하였습니다.

### 결론
1. 하지만, 사용 측면에서 확실히 필터링 조건이 없는 경우 빼고 실행하는게 사실 더 편할 것이라고 판단하였고
2. 필터를 분기 처리가 아닌 쿼리를 통해서 할 수 있는 방법을 찾았습니다.

따라서, 로키가 말씀하신 피드백에 대해서 반영하였습니다. :)

Rok93 commented on src/main/java/roomescape/reservation/ReservationController.java:
```diff
@@ -29,9 +31,10 @@ public ReservationController(
 
     @PostMapping
     public ResponseEntity<ReservationResponse> create(
```
분명 아직 예약가능한 시간대로 떠서 예약을 진행하는데 이미 예약이 되었다는 에러가 발생하고있어요. 👀
어떤 문제가 있을까요? 😃

<img width="1780" alt="image" src="https://github.com/user-attachments/assets/5a2c0798-5a6b-443f-9859-c456acc8bb78" />

changuii replied:
하나의 예약을 식별하는데, < 시간 + 날짜 + 테마 > 가 중복이 되서는 안되지만, < 시간 + 날짜 >로 예약을 검증하고 있었습니다. 반영하였습니다! :)

Rok93 commented on src/main/java/roomescape/auth/JwtProvider.java:
```diff
@@ -0,0 +1,50 @@
+        final Date validity = new Date(now.getTime() + validityInMilliseconds);
+
+        return Jwts.builder()
+                .setClaims(claims)
```
[4단계 미션 페이지의 토큰 생성](https://techcourse.woowahan.com/s/y5znJihX/ls/KINI7JSK) 방식처럼 token에 role 정보도 같이 포함해준다면, 토큰만있다면 별도의 user 정보 조회없이도 인가가 가능할 것 같은데요. 😃 
그렇게된다면 로그인 혹은 권한과 관련된 API를 수행할 때, DB 조회 로직이 하나 줄어들 것 같은데요. 
이 부분에 대해서는 어떻게 생각하시나요?

changuii replied:
보안이라는 요소가 고려하지 않을 수 없는 기준인 것 처럼 느껴져서 트레이드 오프를 생각하며 판단하기가 어려운 것 같습니다. 
로키는 어떤 정보까지 토큰에 담을 수 있다고 생각하시는지 궁금합니다.

### 보안적인 나의 생각
개인적인 생각으로는 ROLE도 탈취한 공격자가 JWT를 디코딩하여, 정보를 확인한다면 ADMIN인 것을 확인하고 매우 악의적으로 활용할 수 있다고 생각이 들었습니다.
- 생각해보니 이메일도 굉장히 개인적인 정보라 고민이 많이 됩니다. 
- 다만 https를 학습해보니 클라이언트(브라우저)와 서버 간에 핸드쉐이킹 과정에서 공개키를 이용하여 대칭키를 교환하고 이 대칭키를 통해 패킷과 메시지를 암호화하여 송수신하기 때문에 인증서 발급 기관을 신뢰한다면 네트워크 채널 상에서 패킷 탈취로 인한 취약성은 걱정하지 않아도 되지 않을까? 라는 생각도 했습니다.

### 결론, 나의 판단
현재 구조가 모든 요청에 대해서 데이터베이스에 쿼리를 사용하고 있어 admin 페이지에 대하여 매 요청마다 `쿼리 비용 + I/O`가 매번 발생하고 이로 인한 단점이 이점보다 훨씬 커지기 때문에 ROLE 정도는 토큰에 담는 것이 더 이점이 크다고 판단하여 반영하였습니다.
- 결국 Session을 사용했을 때 매 요청마다 인증을 다시해주는 것과 같은 비효율이 발생하고, 심지어 메모리가 아닌 데이터베이스라 훨씬 느리다고 판단하였습니다.

Rok93 commented on src/main/java/roomescape/auth/JwtProvider.java:
```diff
@@ -0,0 +1,50 @@
+
+    private final String secretKey = "regjeoigjroigji3j2io3io4h2bjasbdjaksbdkjqu3hu23hru3rhashudhausdhas";
+    private final Key key = Keys.hmacShaKeyFor(Decoders.BASE64.decode(secretKey));
+    private final Long validityInMilliseconds = 60_000L;
```
60s는 토큰 유효시간치고는 너무 짧은 것 같아요. 🥲 

말 나온김에 토큰 유효시간이 길거나 혹은 짧은 경우 각각의 장단점에 대해서 알아보면 어떨까요? 😃

changuii replied:
맞아요 죄송해요 ㅎㅎㅎㅎ,,, 불편하셨을 것 같네요

### 트레이드  오프 지점 판단
토큰의 유효시간이 긴 경우 공개된 채널을 통해 통신하는 토큰이 탈취당했을 경우 공격자가 정당한 사용자처럼 위장하여 토큰을 사용할 수 있다는 문제점이 존재합니다.
따라서, 토큰을 탈취당하더라도 최소한의 피해를 가질 수 있도록 유효시간을 짧게 설정해야합니다.
하지만, 토큰의 시간이 짧다면 사용자가 매번 토큰을 재발급해야하는 불편함이 생기게 됩니다.

따라서 보안과 유저의 편의성간의 트레이드 오프라고 판단하였습니다.

### 나의 생각과 결론
다만 악의적인 공격자가 토큰을 탈취하였을 때 사용자로 사칭하여 공격이 가능하고 사용자의 정보는 매우 민감한 정보라면 이게 트레이드 오프가 가능한 지점인지 의문이 들어 기준을 세우지 못하겠습니다.
- 일반적으로 구글의 API의 경우 토큰의 유효기간은 1시간으로 되어 있는데, 만약 1시간내에 토큰이 탈취된다면 문제가 발생할 수도 있지 않을까? 라는 생각이 듭니다!

일단 유저의 편의성을 고려하여 1시간으로 설정하였습니다.

Rok93 replied:
말씀하신 것처럼 토큰 유효시간이 너무 짧으면 유저의 사용성이 떨어지고 토큰 유효시간이 너무 길게되면 보안의 문제가 발생하게됩니다. 
결국에는 보안을 유지하면서도 사용성을 떨어뜰이지 않는 방식(?)이 이상적이라고 생각할 수 있는데요. 😃

(이미 알고 계실 수도 있겠지만) `refresh token`이란 키워드로 가볍게 알아가면 어떨까 싶어요.

changuii replied:
refresh token을 도입하면 액세스 토큰만 사용했을 때보다 비교적 만료 시간이 짧아지게 되면서 액세스 토큰을 탈취당했을 경우에는, 공격당할 수 있는 시간이 줄어들고, 사용자 경험도 리프레시 토큰의 만료 시간이 길어졌기 때문에, 사용자가 직접 인증을 하는 횟수도 줄어들어 경험도 향상된다고 생각합니다.
다만, 만료 시간이 긴 refreshToken을 만료시간이 되기 전에 처리해버릴 수 있는 장치가 있어야 탈취당했을 경우 대응할 수 있는 방안이 생기기 때문에 이를 위해 토큰 저장소를 구현한다면, 세션과 비슷하게 토큰만을 위한 저장공간이 생긴다는 단점이 생깁니다.
하지만 세션과 다른 가장 큰 차이점은 세션은 매 api 요청마다 인증과 인가를 같이 처리하기 때문에, 매번 데이터베이스에 접근하는 I/O 비용이 가장 크다고 생각하기 때문에 refresh token을 도입하면 비교적 안전(클라이언트 브라우저에서)하고 편리한 사용성을 제공할 수 있다고 생각합니다!

Rok93 reviewed (changes requested):
부기 4~6단계도 잘 구현해주셨네요. 👍💯
질문 주신 부분에 대한 제 생각과 추가 코멘트 남겨두었으니 확인 부탁드릴게요. 😃

추가적으로 IntelliJ IDE로 테스트 수행 시, 테스트 코드가 통과하지 않았던 것은 일시적 현상(?)이었던 것 같아요.  브랜치를 문제의 상항으로 돌려서 테스트 돌렸을 때도 정상적으로 수행하더라구요. 👀

<img width="933" alt="image" src="https://github.com/user-attachments/assets/e79b24ed-7658-4547-9915-95c341052335" />

changuii committed:
feat: 쿠키가 존재하지 않는 경우를 대비해 null체크 추가

changuii committed:
test: 쿠키가 존재하지 않는 경우를 대비해 null체크 테스트

changuii committed:
refactor: Member의 role을 enum으로 관리하도록 변경

changuii committed:
test: Member의 role을 Enum으로 테스트 변경

changuii committed:
feat: 예약을 date, time 중복이 아닌 date, time, theme 중복인 경우 검증으로 변경

changuii committed:
test: 예약을 date, time 중복이 아닌 date, time, theme 중복인 경우 검증으로 테스트 변경

changuii committed:
fix: JWT 토큰 유효기간 60초에서 1시간으로 연장

changuii committed:
refactor: JWT에 역할과 이름 포함 및 TokenBody 도입

- JWT 발급 시 역할(role)과 이름(name) 추가로 포함하도록 수정
- TokenBody 클래스 추가하여 JWT에서 정보 추출 및 관리 간소화
- 회원 검증 로직을 이메일 기반으로 통합 및 검증 메서드 리팩토링
- `LoginMember`에서 ID 제거 및 관련 로직 정리
- Interceptor 및 ArgumentResolver에서 직접 TokenBody 활용
- `email` 컬럼에 UNIQUE 설정 추가

changuii committed:
test: 토큰으로 유저 정보 조회 테스트 제거

- LoginMember에서 불필요한 ID 필드 제거

changuii committed:
feat: favicon.ico 추가

changuii committed:
feat(reservation): 관리자가 예약 필터링 시 AdminFilterReservationRequest DTO 도입 및 관련 로직 수정

관리자 예약 조회 시, DTO를 활용하도록 서비스 및 컨트롤러 수정. SQL 쿼리에 NULL 체크 추가 및 프론트엔드 파라미터명 camelCase로 변경.

changuii committed:
test: AdminFilterReservationRequest 활용으로 테스트 코드 리팩토링

테스트 코드에서 memberId, themeId, 날짜를 개별 매개변수로 전달하던 방식 대신
AdminFilterReservationRequest 객체를 활용하도록 변경하여 가독성과 유지보수성을 개선

Rok93 commented on src/main/java/roomescape/auth/AuthService.java:
```diff
@@ -0,0 +1,46 @@
+        }
+    }
+
+    private void validatePassword(final LoginRequest loginRequest, final Member memberByEmail) {
```
메서드 파라미터의 입장에서는 emain로 알게된 member인지 id로 알게된 member인지에 대해서는 관심사가 아닐 것 같아요. 😃
단순히 `member` 라고 네이밍해도 충분하지 않을까요?

changuii replied:
메서드 추출 과정에서 실수한 것 같습니다. 
말씀하신대로 해당 메서드의 입장에서는 어떤 member가 들어오던지 상관이 없습니다!

Rok93 commented on src/main/java/roomescape/exception/custom/reason/auth/AuthNotExistsCookieException.java:
```diff
@@ -0,0 +1,10 @@
+
+import roomescape.exception.custom.status.BadRequestException;
+
+public class AuthNotExistsCookieException extends BadRequestException {
```
`AuthNotExistsCookieException`는 인증과 관련된 예외인 것 같은데, 400 에러응답을 내려주고있는 것 같아요. 🤔
조금 더 적절한 status code는 없을까요?

changuii replied:
말씀하신 대로, interceptor에서 인가 처리를 하던 중에 토큰이 존재하지 않는다면 401 unauthorization을 응답하는게 인가 처리의 실패를 더 명확하게 제공할 수 있다고 생각합니다! 
현재 구조에서는 resolver와 interceptor에서 공통으로 사용하고 있었기 때문에, 이 예외 들을 분리하여 사용하여 반영했습니다!
- argumentResolver는 인가와 관련된 처리 보다는 토큰에서 값을 가져오는 역할을 하고 있었기 때문에, 400을 응답하도록 하였습니다.

Rok93 reviewed (changes requested):
부기 피드백 반영 잘해주셨네요. 👍👍 
질문 주신 내용에대한 제 생각과 추가 코멘트 남겨두었는데 확인 부탁드릴게요. 🙏
지금도 충분히 잘 구현해주신 것 같아서 다음 리뷰요청 때는 머지하도록할게요. 😉

changuii committed:
refactor: AuthenticationPrincipal 및 관련 클래스 config.auth 패키지로 이동

AuthenticationPrincipal, ArgumentResolver, Interceptor 등의 auth 관련 클래스를 config.auth 패키지로 이동하여 패키지 구조를 개선했습니다. 경로 변경에 따른 import 구문도 수정되었습니다.

changuii committed:
refactor: validatePassword 메서드 파라미터 네이밍 수정

changuii committed:
refactor: 쿠키 부재 예외 처리 수정 및 상태 코드 변경

- AuthNotExistsCookieException을 NotExistsCookieException으로 분리 및 재정의
- 쿠키 부재 시 상태 코드 400 -> 401로 변경
- 테스트 코드와 관련 DisplayName 및 Status Code 수정

Rok93 reviewed (approved):
부기 마지막까지 피드백 반영 잘해주셨어요. 👍💯
방탈출 사용자 예약 미션은 이만 종료하도록하겠습니다. 🎉👏👏 
고생 많으셨고 다음 미션들도 화이팅입니다. 💪💪
